xarray/backends/common.py

```diff
@@ -19,8 +19,12 @@
 if TYPE_CHECKING:
     from io import BufferedIOBase
 
+    from h5netcdf.legacyapi import Dataset as ncDatasetLegacyH5
+    from netCDF4 import Dataset as ncDataset
```

Comment:
Is this a problem if those libraries aren't installed?

Reply:
Since you're asking I'm guessing maybe/probably. I'll find out.

Reply:
So for this particular instance, I'm not sure. it's in the `TYPE_CHECKING` block but mypy didn't seem to care if the libraries were installed or not even with a clean cache. I would have expected a `ModuleNotFoundError` or similiar during mypy checks.

[Here](https://github.com/flamingbear/xarray/blob/mhs%2Fopen_datatree/xarray/backends/h5netcdf_.py#L433) is one place that `h5netcdf` is imported before it is used. My guess is that this is unreachable code if the `h5netcdf` package is not installed by virtue of the `plugins.get_backends(engine)` call**.  Same argument for [netcdf](https://github.com/flamingbear/xarray/blob/mhs%2Fopen_datatree/xarray/backends/netCDF4_.py#L677) and [zarr](https://github.com/flamingbear/xarray/blob/mhs%2Fopen_datatree/xarray/backends/zarr.py#L1044).

This does bring up that the pre-existing functionality may be chagned when opening a datatree and specifying no engine: defaulting to `netcdf4` and falling back to `h5netcdf`. I assumed, maybe incorrectly, that was handled in guessing the engine if not specified.  

So, I think this is ok and welcome new knowlege on wrong I am.


** I only followed that down to `list_engines`' entry_point.backends and now I'm second or third guessing myself.

Reply:
> This does bring up that the pre-existing functionality may be changed when opening a datatree and specifying no engine

Pre-existing in the sense of pre-existing in the xarray-datatree repo? I'm not hugely concerned about that. I'm expecting us to break at least some stuff during the whole merge, and this changing-default isn't very hard for users to correct.

> So, I think this is ok and welcome new knowledge on how wrong I am.

I think unless there is a clear problem (test failure on some environment) then it's fine for now.